feature: mail provider backend

SebastianKrupinski opened the pull request:
* Resolves: #5080

## Summary
Initial Mail Provider Implementation

## TODO
- Copyright
- Comments
- Minor code improvements

SebastianKrupinski committed:
feat: mail provider backend

Signed-off-by: SebastianKrupinski <krupinskis05@gmail.com>

ChristophWurst commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -296,18 +282,53 @@
+			// load mail provider manager
+			$mailManager = \OC::$server->get(\OC\Mail\Provider\Manager::class);
+			// retrieve all services
+			$mailService = $mailManager->findServiceByAddress($this->userSession->getUser()->getUID(), $sender);
```
^

ChristophWurst commented on lib/private/Mail/Provider/Manager.php:
```diff
@@ -0,0 +1,283 @@
+declare(strict_types=1);
+
+/**
+ * @copyright Copyright (c) 2023 Sebastian Krupinski <krupinski01@gmail.com>
```
happy new year. please use spdx license headers

SebastianKrupinski replied:
Lol. Yeah, this was all done before the spdx changes, i will adjust all the files

ChristophWurst commented on lib/private/Mail/Provider/Manager.php:
```diff
@@ -0,0 +1,283 @@
+	}
+
+	/**
+	 * Determain if any mail providers are registered
```
```suggestion
	 * Determine if any mail providers are registered
```

ChristophWurst commented on lib/public/Mail/Provider/IMessageSend.php:
```diff
@@ -0,0 +1,48 @@
+	 * @param IMessage $message			mail message object with all required parameters to send a message
+	 * @param array $options			array of options reserved for future use
+	 */
+	public function messageSend(IMessage $message, array $option = []): void;
```
```suggestion
	public function sendMessage(IMessage $message, array $option = []): void;
```

ChristophWurst reviewed (commented):
Looks sane

SebastianKrupinski commented:
Testing process:

Pull this PR and mail PR - https://github.com/nextcloud/mail/pull/9651

Make sure you have an email account setup in the mail app.

Then create calendar event with an attendee.

Invitation email should be sent from personal mail account

SebastianKrupinski commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -268,18 +255,57 @@ public function schedule(Message $iTipMessage) {
+				$mailService = $mailManager->findServiceByAddress($this->userSession->getUser()->getUID(), $sender);
+			}
+			// evaluate if a mail service was found and has sending capabilities
+			if ($mailService !== null && $mailService instanceof IMessageSend) {
```
I left the check for null in place, since the instanceOf check triggers the class loader to find and load the class file, its more processing efficient to check for null first.

SebastianKrupinski commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -268,18 +255,57 @@ public function schedule(Message $iTipMessage) {
+				$mailService = $mailManager->findServiceByAddress($this->userSession->getUser()->getUID(), $sender);
+			}
+			// evaluate if a mail service was found and has sending capabilities
+			if ($mailService !== null && $mailService instanceof IMessageSend) {
```
Done

SebastianKrupinski commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -268,18 +255,57 @@ public function schedule(Message $iTipMessage) {
+				// load mail provider manager
+				$mailManager = \OC::$server->get(MailManager::class);
+				// retrieve all services
+				$mailService = $mailManager->findServiceByAddress($this->userSession->getUser()->getUID(), $sender);
```
Done

SebastianKrupinski commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -268,18 +255,57 @@
+				// load mail provider manager
+				$mailManager = \OC::$server->get(MailManager::class);
+				// retrieve all services
+				$mailService = $mailManager->findServiceByAddress($this->userSession->getUser()->getUID(), $sender);
```
Done

SebastianKrupinski commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -268,18 +255,57 @@ public function schedule(Message $iTipMessage) {
 		try {
-			$failed = $this->mailer->send($message);
+			// evaluate if user object exists
+			if ($this->userSession->getUser() !== null) {
```
Done

SebastianKrupinski commented on lib/public/Mail/Provider/IManager.php:
```diff
@@ -0,0 +1,112 @@
+	 * @since 30.0.0
+	 *
+	 */
+	public function register(IProvider $value): void;
```
True, at the moment all registration take place through IRegistrationContext. This was just a method I saw in the other mangers I looked at and didn't know if it was necessary. It was useful, in live standalone testing tho, which allowed me to test everything without loading all the apps. Not sure if it might be useful to keep.

kesselb commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -268,18 +254,65 @@
-			$failed = $this->mailer->send($message);
+			// evaluate if user object exists
+			if ($this->userSession->getUser() !== null) {
+				$userId = $this->userSession->getUser()->getUID();
```
Psalm is still unhappy :see_no_evil: 

`getUser` returns an IUser instance or null, and therefore we need to check the return value. 
`getUID` does not return null. 

Change it like below to make psalm happy: 

```php
// evaluate if user object exists
$user = $this->userSession->getUser();
// evaluate if user id exists
if ($user !== null) {
	// load mail provider manager
	$mailManager = \OCP\Server::get(\OCP\Mail\Provider\IManager::class);
	// retrieve appropriate service with the same address as sender
	$mailService = $mailManager->findServiceByAddress($user->getUID(), $sender);
}
```

SebastianKrupinski replied:
Yeah I noticed this after I pushed the changes. I've applied your recommendations.

kesselb commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -268,18 +255,57 @@ public function schedule(Message $iTipMessage) {
+			// evaluate if user object exists
+			if ($this->userSession->getUser() !== null) {
+				// load mail provider manager
+				$mailManager = \OC::$server->get(MailManager::class);
```
Can you try to inject the manager instance via dependency injection?

SebastianKrupinski replied:
Yes. I can add it to the constructor, just means I will have to update a bunch more files because we manually load this class in several places. I was also trying to avoid loading extra classes on creation as this will only get used when a event has an attendee. But DI would make it work better with phpStorm.

SebastianKrupinski replied:
Done

kesselb commented on lib/public/Mail/Provider/IService.php:
```diff
@@ -0,0 +1,141 @@
+	 *
+	 * @return bool|array					true/false if ability is supplied, collection of abilities otherwise
+	 */
+	public function capable(?string $ability = null): bool | array;
```
Please split in one method to check if a service could provide a given ability and another one to return a collection of abilities.

st3iny replied:
I agree. Have one method `getCapabilities()` or `capabilities()` and one `isCapable()` to check a single one.

SebastianKrupinski replied:
I wanted to avoid using class global variables and containing the code in one method for simplicity. But I will break this up. I also just recently realized that different services might actually not support the same capabilities even tho the use the same protocol.

kesselb commented on apps/dav/tests/unit/CalDAV/Schedule/IMipPluginTest.php:
```diff
@@ -107,6 +111,8 @@ protected function setUp(): void {
 
 		$this->eventComparisonService = $this->createMock(EventComparisonService::class);
 
+		$this->mailManager = $this->createMock(IMailManager::class);
```
Some basic tests, that the code path with the mail manager also works, would be nice.

kesselb commented on lib/public/Mail/Provider/IManager.php:
```diff
@@ -0,0 +1,104 @@
+	 *
+	 * @return array<string,String>
+	 */
+	public function types(): array;
```
What's the use case for the types method, and why is it called types?

It returns [$providerId => $providerLabel], why is that a type?

st3iny replied:
Again nitpicking here: How about calling it `names()`?

SebastianKrupinski replied:
Its called types because it returns the type of providers available "IMAP/JMAP/EWS/etc", originally it only returned [0 => 'IMAP'] but I realized that it could also be used in a Mail Setup wizard scenario, in which case the human friendly name "IMAP Provider" would be useful, so i changed to  ['type' => 'label'] e.g. ['IMAP' => 'IMAP Provider'] then an app can list all the provider types available and display them. The purpose is NOT to provide a name.

kesselb commented on lib/public/Mail/Provider/IProvider.php:
```diff
@@ -0,0 +1,117 @@
+	public function id(): string;
+
+	/**
+	 * The localized human frendly name of this provider
```
```suggestion
	 * The localized human friendly name of this provider
```

SebastianKrupinski replied:
Done

kesselb commented on lib/public/Mail/Provider/IProvider.php:
```diff
@@ -0,0 +1,117 @@
+	public function label(): string;
+
+	/**
+	 * Determain if any services are configured for a specific user
```
```suggestion
	 * Determine if any services are configured for a specific user
```

SebastianKrupinski replied:
Done

kesselb commented on lib/public/Mail/Provider/IProvider.php:
```diff
@@ -0,0 +1,117 @@
+	public function findServiceByAddress(string $uid, string $address): IService | null;
+
+	/**
+	 * create a service configuration for a specific user
```
```suggestion
	 * Create a service configuration for a specific user
```

SebastianKrupinski replied:
Done

kesselb commented on lib/public/Mail/Provider/IProvider.php:
```diff
@@ -0,0 +1,117 @@
+	 * @param string $uid				user id
+	 * @param string $address			mail address (e.g. test@example.com)
+	 *
+	 * @return IService					returns service object or null if non found
```
```suggestion
	 * @return ?IService					returns service object or null if non found
```

I wish the method would rather throw a ServiceNotFoundException than returning null. Some people argue that exceptions are bad for performance and still, I find it much cleaner to work with exceptions than all those !== null checks.

SebastianKrupinski replied:
This method is used by the manager to find services as it itereates through providers, for both performance and code complexity its simpler to just return a null, instead of handling errors.

kesselb commented on lib/public/Mail/Provider/IService.php:
```diff
@@ -0,0 +1,141 @@
+	 *
+	 * @return array<int, IAddress>			collection of mail address objects
+	 */
+	public function getSecondaryAddress(): array | null;
```
```suggestion
	public function getSecondaryAddress(): array | null;
```

According to the phpdoc it returns a collection and therefore please rename to getSecondaryAddresses and change the return type to array.

SebastianKrupinski replied:
Done

kesselb commented on tests/lib/Mail/Provider/ManagerTest.php:
```diff
@@ -0,0 +1,189 @@
+
+class ManagerTest extends TestCase {
+
+	/** @var CoordinatorMockObject*/
+	private $coordinator;
```
```suggestion
	private Coordinator&MockObject $coordinator;
```

I usually omit &MockObject because PHPStorm still provides me the proper code completion for the mock object but that's your call. However `CoordinatorMockObject` does not work ;)

The same for $container, $logger, $provider and $service.

SebastianKrupinski replied:
I must have accidentally clicked resolve on the wrong comment!

SebastianKrupinski commented on lib/private/Mail/Provider/Manager.php:
```diff
@@ -0,0 +1,253 @@
+	 *
+	 * @return IService|null			returns service object or null if non found
+	 */
+	public function findServiceById(string $uid, string $sid, ?string $pid = null): IService | null {
```
Done

SebastianKrupinski commented on lib/private/Mail/Provider/Manager.php:
```diff
@@ -0,0 +1,253 @@
+	public function findServiceById(string $uid, string $sid, ?string $pid = null): IService | null {
+		
+		// evaluate if provider id was specified
+		if ($pid !== null) {
```
Yes, the consumer would store both the service id and the provider id (best case), but they could also store just the service id if the id is unique like a uuid.

SebastianKrupinski commented on lib/private/Mail/Provider/Manager.php:
```diff
@@ -0,0 +1,253 @@
+		// evaluate if provider id was specified
+		if ($pid !== null) {
+			// find provider
+			$provider = $this->findProviderById($pid);
```
I would perfer NOT to throw errors, for instance if let's say the mail app was loading multiple accounts with different service id and provider id and one of the providers was temporarly disabled that would cause the entire process to fail.

SebastianKrupinski commented on lib/private/Mail/Provider/Manager.php:
```diff
@@ -0,0 +1,252 @@
+				// add provider to cache collection
+				$this->providersCollection[$provider->id()] = $provider;
+			} catch (Throwable $e) {
+				$this->logger->error(
```
fixed

SebastianKrupinski commented on lib/public/Mail/Provider/Message.php:
```diff
@@ -0,0 +1,344 @@
+		// initilize data store
+		$this->data = [];
+		// evaluate if id was passed otherwise generate a new id
+		$this->data['id'] = isset($id) ? $id : bin2hex(random_bytes(16));
```
We don't require the id for all functions, for instance creating or sending a new message does not require an id. And different protocols/systems (IMAP/JMAP/EWS) have different id schemes and asign there on ids.

SebastianKrupinski replied:
Actually I've been thinking about instancing the message from the provider/service it self, that way a provider/service can create a custom message object with custom logic for each method and additional methods if required.

SebastianKrupinski commented on lib/public/Mail/Provider/Message.php:
```diff
@@ -0,0 +1,344 @@
+ */
+class Message implements \OCP\Mail\Provider\IMessage {
+
+	protected array $data;
```
Readonly properties - As of PHP 8.1.0, a property can be declared with the readonly modifier, which prevents modification of the property after initialization. 

This would mean that this varialbe is the same as a static variable and can not be changed. We change the data inside of this variable with the methods. Its protected so that it can be directly accessed if this class it extended.

kesselb replied:
I've suggested using readonly properties because IMessage looks like a DTO object. Starting with PHP 8.1 we can use immutable DTOs with just a few lines of code: https://stitcher.io/blog/php-81-readonly-properties

st3iny replied:
For reference, here is what kesselb is talking about:

```php
class BlogData
{
    public function __construct(
        public readonly string $title,
        public readonly Status $status,
        public readonly ?DateTimeImmutable $publishedAt = null,
    ) {}
}
```

SebastianKrupinski replied:
Read only properties will not work in this case. This property needs to be altered after the constructor.

SebastianKrupinski commented on lib/public/Mail/Provider/Message.php:
```diff
@@ -0,0 +1,344 @@
+	 *
+	 * @since 30.0.0
+	 *
+	 * @param array<int,IAttachment>		collection of all mail attachment objects
```
Done

SebastianKrupinski commented on lib/public/Mail/Provider/IMessage.php:
```diff
@@ -0,0 +1,232 @@
+	 *
+	 * @since 30.0.0
+	 *
+	 * @param array<int,IAttachment>			collection of all mail attachment objects
```
Done

SebastianKrupinski commented on lib/public/Mail/Provider/IMessageSend.php:
```diff
@@ -0,0 +1,31 @@
+	 * @since 30.0.0
+	 *
+	 * @param IMessage $message			mail message object with all required parameters to send a message
+	 * @param array $options			array of options reserved for future use
```
If you are talking about the $options... One of the uses would be to override any caching that the provider might implement, simple ['CACHE' = false] another one might be encryption signing keys, basically anything that might need to be implemented that I havent thought of yet. Lol. API's take a long time to change once they are implemented, so we might need something tempory to make a feature work.

st3iny replied:
I've discussed this with Sebastian in Hamburg and would prefer to leave it out but don't want to block this PR any further. 

In the future we could always add optional parameters to the function. And we could always add another method with a different set of parameters to prevent breaking compatibility (like I did with the public dashboard APIs a while back).

SebastianKrupinski commented on lib/private/Mail/Provider/Manager.php:
```diff
@@ -0,0 +1,253 @@
+			}
+		}
+		// return mail provider collection
+		return $this->providersCollection;
```
This code generate a nested array to providers and services under each provider. So there should be no issue.

```
[
	'Provider1' => [
		['Service1' => ServiceObject],
		['Service2' => ServiceObject]
	],
	'Provider2' => [
		['Service1' => ServiceObject],
		['Service2' => ServiceObject]
	],
]
```

kesselb commented:
- IAddress.getAddress
- IAddress.getLabel
- IAttachment.getName
- IAttachment.getType
- IAttachment.getContents
- IMessage.getFrom
- IMessage.getReplyTo
- IMessage.getTo
- IMessage.getCC
- IMessage.getBcc
- IMessage.getSubject
- IMessage.getBody
- IMessage.getBodyHtml
- IMessage.getBodyPlain
- IMessage.getAttachments

Each of the above methods can return null. 
This should be reworked. 

For example, IAttachment: 

What's the use case for having an attachment object with a name and a type but content's null, and what is the consumer supposed to do with such an object?

- https://github.com/nextcloud/mail/pull/9651/files#diff-396c7c8ae7dacf5cd102f6d592a4dc8b485833c22e9265d4b24b062164d2db72R47-R52 
- https://github.com/nextcloud/mail/blob/dd160e64d854ef915f4eab9217b1c84b9902bfa4/lib/Service/Attachment/AttachmentService.php#L95

![Screenshot from 2024-07-15 12-32-09](https://github.com/user-attachments/assets/2c1525f2-fe61-45cc-bff4-f43c6ff600cf)

![Screenshot from 2024-07-15 12-32-27](https://github.com/user-attachments/assets/77c2872d-75ef-4714-a5d7-d5e885b3200a)

```
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>TypeError</s:exception>
  <s:message>OCA\Mail\Service\Attachment\AttachmentService::addFileFromString(): Argument #4 ($fileContents) must be of type string, null given, called in /var/www/html/apps-extra/mail/lib/Provider/Command/MessageSend.php on line 47</s:message>
</d:error>
```



The requirements for the public api should be as strict as possible and can be extended when necessary. We don't have a use case for having a "null" attachment and therefore should not implement it. The same for every other method from the list above.

kesselb commented:
- IProvider.createService
- IProvider.modifyService
- IProvider.deleteService

Are unused and the implemention in https://github.com/nextcloud/mail/pull/9651/files#diff-863bf0fae4943be717ec0db9eba3ff621d28665362e192e0b6d45a2b917426f7R190-R236 does nothing and therefore should be removed.

SebastianKrupinski commented on tests/lib/Mail/Provider/ManagerTest.php:
```diff
@@ -0,0 +1,189 @@
+	private $coordinator;
+	/** @var ContainerInterfaceMockObject*/
+	private $container;
+	/** @var LoggerInterfaceMockObject*/
```
Done

SebastianKrupinski commented on tests/lib/Mail/Provider/ManagerTest.php:
```diff
@@ -0,0 +1,189 @@
+	private $logger;
+	/** @var IProviderMockObject*/
+	private $provider;
+	/** @var IServiceMockObject*/
```
Done

st3iny commented on lib/public/Mail/Provider/IMessage.php:
```diff
@@ -0,0 +1,232 @@
+	 */
+	public function getSubject(): string | null;
+
+	/**
+	 * sets the plain text or html body of this message
+	 *
+	 * @since 30.0.0
+	 *
+	 * @param string $value                     text or html body of message
+	 * @param bool $html						html flag - true for html
+	 *
+	 * @return self                             return this object for command chaining
+	 */
+	public function setBody(string $value, bool $html): self;
+
+	/**
+	 * gets either the html or plain text body of this message
+	 *
+	 * html body will be returned over plain text if html body exists
+	 *
+	 * @since 30.0.0
+	 *
+	 * @param string|null                       html/plain body of this message or null if one is not set
+	 */
+	public function getBody(): string | null;
```
Yeah, I agree. I'm thinking about applications were both might be needed.

SebastianKrupinski replied:
Email supports both html and plain text bodies there for both need to be implemented. This is the case for IMAP/JMAP/EWS.

st3iny reviewed (commented):
The API looks solid to me and I like the design overall.

However, I agree with kesselb's latest points and have some additional minor high level feedback:
1. Please get rid of `IServiceIdentity` and `IServiceLocation` interfaces. We can always add new interfaces once those two features are needed and ready. Currently, they seem to be unused.
2. All methods inside `IMessage` that return an `array` or `null` should only return an (empty) `array`. This makes consuming the API more convenient as both cases don't need to be handled separately despite being semantically identical. (As kesselb mentioned, there are some more methods inside `IAttachment` where this also applies.)

SebastianKrupinski commented on tests/lib/Mail/Provider/ManagerTest.php:
```diff
@@ -0,0 +1,189 @@
+	private $container;
+	/** @var LoggerInterfaceMockObject*/
+	private $logger;
+	/** @var IProviderMockObject*/
```
Done

SebastianKrupinski commented:
@st3iny @kesselb 

I've changed the empty return type on the functions that return an array to an empty array, you have a valid point about the easier consumption of the array.

I have also removed the extra IServiceIdentity and IServiceLocation interfaces and methods, even tho I have a feeling I might be adding them back in next month from a conversion I had this week, but I save the code so that should not be a big deal.

miaulalala reviewed (approved):
Aside from the comments by Daniel and Richard this looks fine.

st3iny reviewed (approved):
Feedback has been addressed.

kesselb reviewed (dismissed):
https://github.com/nextcloud/server/pull/45383#issuecomment-2228196994

kesselb commented:
I'm aware that Beta 1 is due on 2024-07-25, but we can always backport changes if necessary.

- [ ] Most methods still accept null values for no reason. I would prefer to sort that out before merging. It's easier to make an API less strict later, than the other way around. 
- [ ] https://github.com/nextcloud/server/blob/cf432cf1c31908a45adde486ff4353a37254e8ed/lib/public/Mail/Provider/Message.php Please use class properties rather then `$this->data`.

kesselb commented:
Sebastian and I discussed the nullable return types and agreed to keep the nullable return types.

ChristophWurst commented:
@SebastianKrupinski this PR adds new APIs to lib/public aka OCP, so we need developer docs. You can take https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/groupware/calendar.html as inspiration.

susnux commented:
Documentation was added: https://github.com/nextcloud/documentation/pull/12083
